lib/Backends/OpenCL/kernels.cl

```diff
@@ -1056,54 +1056,40 @@ __kernel void oclpoolavgW(__global void *mem, cl_uint32_t dest, cl_uint32_t src,
+/// \p type the type of the tensor elements and of the return value
+#define DEFINE_OPENCL_TRANSPOSE_KERNEL(name, type)                             \
+  __kernel void transpose##name##K(__global type *dest, __global type *src,    \
+                                   ShapeNHWC odim, ShapeNHWC idim,             \
+                                   ShapeNHWC shuffle) {                        \
+    size_t res[4];                                                             \
+    size_t d0 = get_global_id(0);                                              \
+    size_t d1 = get_global_id(1);                                              \
+    res[0] = d0;                                                               \
+    res[1] = d1;                                                               \
+    for (size_t d2 = 0; d2 < idim.w; d2++) {                                   \
+      res[2] = d2;                                                             \
+      for (size_t d3 = 0; d3 < idim.c; d3++) {                                 \
+        res[3] = d3;                                                           \
+        size_t dstIdx = getNHWC(odim, res[shuffle.n], res[shuffle.h],          \
+                                res[shuffle.w], res[shuffle.c]);               \
+        size_t srcIdx = getNHWC(idim, d0, d1, d2, d3);                         \
+        dest[dstIdx] = src[srcIdx];                                            \
+      }                                                                        \
+    }                                                                          \
+  }                                                                            \
+  __kernel void transpose##name##W(__global void *mem, cl_uint32_t dest,       \
+                                   cl_uint32_t src, ShapeNHWC odim,            \
+                                   ShapeNHWC idim, ShapeNHWC shuffle) {        \
+    transpose##name##K(&mem[dest], &mem[src], odim, idim, shuffle);            \
   }
-}
 
-__kernel void transposeW(__global void *mem, cl_uint32_t dest, cl_uint32_t src,
-                         ShapeNHWC odim, ShapeNHWC idim, ShapeNHWC shuffle) {
-  transposeK(&mem[dest], &mem[src], odim, idim, shuffle);
-}
-
-__kernel void transposeK_u(__global cl_uint64_t *dest,
-                           __global cl_uint64_t *src, ShapeNHWC odim,
-                           ShapeNHWC idim, ShapeNHWC shuffle) {
-  size_t res[4];
-  size_t d0 = get_global_id(0);
-  size_t d1 = get_global_id(1);
-  res[0] = d0;
-  res[1] = d1;
-  for (size_t d2 = 0; d2 < idim.w; d2++) {
-    res[2] = d2;
-    for (size_t d3 = 0; d3 < idim.c; d3++) {
-      res[3] = d3;
-      size_t dstIdx = getNHWC(odim, res[shuffle.n], res[shuffle.h],
-                              res[shuffle.w], res[shuffle.c]);
-      size_t srcIdx = getNHWC(idim, d0, d1, d2, d3);
-      dest[dstIdx] = src[srcIdx];
-    }
-  }
-}
+DEFINE_OPENCL_TRANSPOSE_KERNEL(_i8, cl_int8_t)
```

Comment:
either pass a full name here or name the parameter of the macro `suffix` instead of `name`. Otherwise it looks a bit odd. I'd prefer passing a full name here, because `_i8` looks strange.

Reply:
make sense. Changed to pass full name here.